src/Schedule.cpp

```diff
@@ -36,8 +40,64 @@ EXPORT void destroy<ScheduleContents>(const ScheduleContents *p) {
     delete p;
 }
 
+namespace {
+
+// Deep-copy data from 'src' to 'dst'. Ignore copying 'specialization' over.
+void deep_copy_schedule_contents_helper(
+        IntrusivePtr<ScheduleContents> &dst, const IntrusivePtr<ScheduleContents> &src,
+        std::map<IntrusivePtr<FunctionContents>, IntrusivePtr<FunctionContents>> &copied_map) {
+
+    if (!src.defined()) {
+        dst = src;
+        return;
+    }
+    dst = IntrusivePtr<ScheduleContents>(new ScheduleContents);
+    dst.ptr->store_level = src.ptr->store_level;
+    dst.ptr->compute_level = src.ptr->compute_level;
+    dst.ptr->splits = src.ptr->splits;
+    dst.ptr->dims = src.ptr->dims;
+    dst.ptr->storage_dims = src.ptr->storage_dims;
+    dst.ptr->bounds = src.ptr->bounds;
+    dst.ptr->reduction_domain = src.ptr->reduction_domain.deep_copy();
+    dst.ptr->memoized = src.ptr->memoized;
+    dst.ptr->touched = src.ptr->touched;
+    dst.ptr->allow_race_conditions = src.ptr->allow_race_conditions;
+
+    // Deep-copy wrapper functions
+    for (const auto &iter : src.ptr->wrappers) {
+        if (copied_map.count(iter.second)) {
+            dst.ptr->wrappers[iter.first] = copied_map[iter.second];
+        } else {
+            dst.ptr->wrappers[iter.first] = deep_copy_function_contents_helper(iter.second, copied_map);
+            copied_map[iter.second] = dst.ptr->wrappers[iter.first];
+        }
+    }
+    internal_assert(dst.ptr->wrappers.size() == src.ptr->wrappers.size());
+}
+
+}
+
 Schedule::Schedule() : contents(new ScheduleContents) {}
 
+void Schedule::deep_copy(
+        Schedule &copy,
+        std::map<IntrusivePtr<FunctionContents>, IntrusivePtr<FunctionContents>> &copied_map) const {
```

Comment:
This map type should probably go in a typedef, especially if it's used anywhere else...